dot/sync/disjoint_block_set.go

```diff
@@ -90,6 +113,33 @@ func newDisjointBlockSet(limit int) *disjointBlockSet {
 	}
 }
 
+func (s *disjointBlockSet) start(ctx context.Context) {
```

Comment:
A suggestion for here and other 'start'-like functions.

The problem with these 'fire-and-forget' start goroutines is that we don't know when they are done.
The simplest way to change this is for `start` to return a `done <-chan struct{}` channel which is closed by a `defer close(done)` inside the goroutine launched. The caller should then wait (or ignore) when e.g. canceling the context for the goroutine to cleanly exit.

Ideally, to also reduce indentation and complexity of the function, change it to a

```go
func (s *disjointBlockSet) run(ctx context.Context, done chan<- struct{}) {
```

Which would run the for loop directly without lauching a goroutine. The caller would have to use `go dBlockSet.run(ctx, done)` to run async if he wants to. That way it's up to the caller to run it in a blocking manner or not.

Finally, you can also change `done` to be an error channel if you want to get a final error such as `ctx.Err()` or some other error.

Reply:
I have seen in some places like in the Docker repo, where there is also a struct channel passed in that is closed right before the `select` to signal the caller the goroutine is ready and operational. I don't really do that personally but it could be nice :wink:

Reply:
updated!